rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -658,6 +690,34 @@ void Player::create_control_services()
     {
       response->success = play_next();
     });
+  srv_play_for_ = create_service<rosbag2_interfaces::srv::PlayFor>(
+    "~/play_for",
+    [this](
+      const std::shared_ptr<rmw_request_id_t>/* request_header */,
+      const std::shared_ptr<rosbag2_interfaces::srv::PlayFor::Request> request,
+      const std::shared_ptr<rosbag2_interfaces::srv::PlayFor::Response> response)
+    {
+      const rcutils_duration_value_t duration =
+      static_cast<rcutils_duration_value_t>(request->duration.sec) *
+      static_cast<rcutils_duration_value_t>(1000000000) +
+      static_cast<rcutils_duration_value_t>(request->duration.nanosec);
+      play({duration});
+      response->success = true;
+    });
+  srv_play_until_ = create_service<rosbag2_interfaces::srv::PlayUntil>(
+    "~/play_until",
+    [this](
+      const std::shared_ptr<rmw_request_id_t>/* request_header */,
+      const std::shared_ptr<rosbag2_interfaces::srv::PlayUntil::Request> request,
+      const std::shared_ptr<rosbag2_interfaces::srv::PlayUntil::Response> response)
+    {
+      const rcutils_time_point_value_t timestamp =
+      static_cast<rcutils_time_point_value_t>(request->time.sec) *
+      static_cast<rcutils_time_point_value_t>(1000000000) +
+      static_cast<rcutils_time_point_value_t>(request->time.nanosec);
```

Comment:
@gbiggs nit: consider
```suggestion
      const rcutils_time_point_value_t timestamp =
          rclcpp::Time(request->time).nanoseconds();
```